File: Deprerated/Siren/Attribute/SirenFieldAttribute.cs
// Copyright (c) 2015 fjz13. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.
using System;

namespace Siren.Attribute
{

    public enum SirenFieldModifier
    {
        Optional = 0,
        Required = 1,
    };


    [AttributeUsage(AttributeTargets.Property)]
    public class SirenFieldAttribute : System.Attribute
    {
        public SirenFieldModifier Modifier { get; set; }
        public string NewName { get; set; }
        public object DefaultValue { get; set; }
        public bool ForceKeyToPtr { get; set; }
        public bool ForceValueToPtr { get; set; }
        public bool AddDictionaryMethods { get; set; }
        public bool SuppressMethod { get; set; }




        public SirenFieldAttribute(SirenFieldModifier modifier = SirenFieldModifier.Required,object defaultValue=null)
        {
            Modifier = modifier;
            DefaultValue = defaultValue;
        }
      
    }
}

File: Medusa/MedusaCore/Core/Siren/Schema/Attribute/SirenFieldAttribute.cpp
// Copyright (c) 2015 fjz13. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.
#include "MedusaCorePreCompiled.h"
#include "SirenFieldAttribute.h"
#include "Core/IO/Stream/IStream.h"
#include "Core/Log/Log.h"

MEDUSA_BEGIN;


SirenFieldAttribute::~SirenFieldAttribute(void)
{

}


bool SirenFieldAttribute::IsRequired() const
{
	return !MEDUSA_FLAG_HAS(mMode, SirenFieldGenerateMode::Optional);
}


bool SirenFieldAttribute::OnLoaded()
{
	StringPropertySet copy = mKeyValues;
	if (mKeyValues.RemoveKey("Optional"))
	{
		MEDUSA_FLAG_ADD(mMode, SirenFieldGenerateMode::Optional);
	}

	if (mKeyValues.RemoveKey("?"))
	{
		MEDUSA_FLAG_ADD(mMode, SirenFieldGenerateMode::Optional);
	}

	if (mKeyValues.RemoveKey("+"))
	{
		MEDUSA_FLAG_REMOVE(mMode, SirenFieldGenerateMode::Optional);
	}

	if (mKeyValues.RemoveKey("ForceKeyToPtr"))
	{
		MEDUSA_FLAG_ADD(mMode, SirenFieldGenerateMode::ForceKeyToPtr);
	}

	if (mKeyValues.RemoveKey("ForceValueToPtr"))
	{
		MEDUSA_FLAG_ADD(mMode, SirenFieldGenerateMode::ForceValueToPtr);
	}

	if (mKeyValues.RemoveKey("AddDictionaryMethods"))
	{
		MEDUSA_FLAG_ADD(mMode, SirenFieldGenerateMode::AddDictionaryMethods);
	}

	if (mKeyValues.RemoveKey("SuppressMethod"))
	{
		MEDUSA_FLAG_ADD(mMode, SirenFieldGenerateMode::SuppressMethod);
	}


	return true;
}

StringRef SirenFieldAttribute::Modifier() const
{
	if (IsRequired())
	{
		return "Required";
	}
	return "Optional";
}

bool SirenFieldAttribute::LoadFrom(IStream& stream)
{
	RETURN_FALSE_IF_FALSE(ISirenAttribute::LoadFrom(stream));
	mMode = stream.Read<SirenFieldGenerateMode>();
	return true;
}

bool SirenFieldAttribute::SaveTo(IStream& stream) const
{
	RETURN_FALSE_IF_FALSE(ISirenAttribute::SaveTo(stream));
	stream.Write(mMode);
	return true;
}

MEDUSA_END;